Python/jit.c

```diff
@@ -146,256 +144,257 @@ set_bits(uint32_t *loc, uint8_t loc_start, uint64_t value, uint8_t value_start,
+    // There might be an implicit shift encoded in the instruction:
+    uint8_t shift = 0;
+    if (IS_AARCH64_LDR_OR_STR(*loc32)) {
+        shift = (uint8_t)get_bits(*loc32, 30, 2);
+        // If both of these are set, the shift is supposed to be 4.
+        // That's pretty weird, and it's never actually been observed...
+        assert(get_bits(*loc32, 23, 1) == 0 || get_bits(*loc32, 26, 1) == 0);
     }
+    value = get_bits(value, 0, 12);
+    assert(get_bits(value, 0, shift) == 0);
+    set_bits(loc32, 10, value, shift, 12);
 }
 
-static void
-copy_and_patch(unsigned char *base, const Stencil *stencil, uintptr_t patches[])
+void
+patch_aarch64_12x(unsigned char *location, uint64_t value)
 {
-    memcpy(base, stencil->body, stencil->body_size);
-    patch(base, stencil, patches);
+    patch_aarch64_12(location, value);
 }
 
-static void
-emit(const StencilGroup *group, uintptr_t patches[])
+// 16-bit low part of an absolute address.
+void
+patch_aarch64_16a(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    assert(IS_AARCH64_MOV(*loc32));
+    // Check the implicit shift (this is "part 0 of 3"):
+    assert(get_bits(*loc32, 21, 2) == 0);
+    set_bits(loc32, 5, value, 0, 16);
+}
+
+// 16-bit middle-low part of an absolute address.
+void
+patch_aarch64_16b(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    assert(IS_AARCH64_MOV(*loc32));
+    // Check the implicit shift (this is "part 1 of 3"):
+    assert(get_bits(*loc32, 21, 2) == 1);
+    set_bits(loc32, 5, value, 16, 16);
+}
+
+// 16-bit middle-high part of an absolute address.
+void
+patch_aarch64_16c(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    assert(IS_AARCH64_MOV(*loc32));
+    // Check the implicit shift (this is "part 2 of 3"):
+    assert(get_bits(*loc32, 21, 2) == 2);
+    set_bits(loc32, 5, value, 32, 16);
+}
+
+// 16-bit high part of an absolute address.
+void
+patch_aarch64_16d(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    assert(IS_AARCH64_MOV(*loc32));
+    // Check the implicit shift (this is "part 3 of 3"):
+    assert(get_bits(*loc32, 21, 2) == 3);
+    set_bits(loc32, 5, value, 48, 16);
+}
+
+// 21-bit count of pages between this page and an absolute address's page... I
+// know, I know, it's weird. Pairs nicely with patch_aarch64_12 (above).
+void
+patch_aarch64_21r(unsigned char *location, uint64_t value)
+{
+    uint32_t *loc32 = (uint32_t *)location;
+    value = (value >> 12) - ((uintptr_t)location >> 12);
+    // Check that we're not out of range of 21 signed bits:
+    assert((int64_t)value >= -(1 << 20));
+    assert((int64_t)value < (1 << 20));
+    // value[0:2] goes in loc[29:31]:
+    set_bits(loc32, 29, value, 0, 2);
+    // value[2:21] goes in loc[5:26]:
+    set_bits(loc32, 5, value, 2, 19);
+}
+
+void
+patch_aarch64_21rx(unsigned char *location, uint64_t value)
```

Comment:
I guess same comment here.

Reply:
That said, I'm trying to understand the function name suffixes. At least, `r` suffixed name seems to mean relative addressing. What's `rx`?

I guess meta comment but is it worth adding a comment at the top of the block before defining the functions about the convention, just for clarity's sake?

Reply:
Yeah, maybe I should expand the cryptic names.

`r` is "relative", and `x` is "relax" (meaning, try to rewrite the instruction into something more efficient). Naming is hard, especially when two words you want to include in the name start with the same three letters. 🙃

`patch_aarch64_21rx` and `patch_aarch64_12x` are sort of special, since they can't be relaxed individually, but they *can* be relaxed as a pair. That's why we need `Hole.fold`: if it sees a pair of these, it'll turn them into a special `patch_aarch64_33rx` call. Otherwise, they don't do anything different than their non-relaxing versions.

Reply:
Ah okay, thanks for explaining. That makes sense. Probably just a comment somewhere should suffice?